apollo-router/src/services/layers/apq.rs

```diff
@@ -34,101 +34,137 @@ struct PersistedQuery {
+            cache: self.cache.clone(),
+            inner: Buffer::new(service, DEFAULT_BUFFER_SIZE),
+        }
+    }
+}
+
+pub(crate) struct APQService<S>
+where
+    S: Service<RouterRequest>,
+{
+    cache: DeduplicatingCache<Vec<u8>, String>,
+    inner: Buffer<S, RouterRequest>,
+}
+
+impl<S> Service<RouterRequest> for APQService<S>
+where
+    S: Service<
+            RouterRequest,
+            Response = RouterResponse<BoxStream<'static, Response>>,
+            Error = BoxError,
+        > + Send
+        + 'static,
+    <S as Service<RouterRequest>>::Future: Send + 'static,
+{
+    type Response = RouterResponse<BoxStream<'static, Response>>;
+
+    type Error = BoxError;
+
+    type Future =
+        BoxFuture<'static, Result<RouterResponse<BoxStream<'static, Response>>, BoxError>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: RouterRequest) -> Self::Future {
+        let maybe_query_hash: Option<Vec<u8>> = req
+            .originating_request
+            .body()
+            .extensions
+            .get("persistedQuery")
+            .and_then(|value| serde_json_bytes::from_value::<PersistedQuery>(value.clone()).ok())
+            .and_then(|persisted_query| hex::decode(persisted_query.sha256hash.as_bytes()).ok());
+
+        let body_query = req.originating_request.body().query.clone();
+
         let cache = self.cache.clone();
-        CheckpointService::new(
-            move |mut req| {
-                let maybe_query_hash: Option<Vec<u8>> = req
-                    .originating_request
-                    .body()
-                    .extensions
-                    .get("persistedQuery")
-                    .and_then(|value| {
-                        serde_json_bytes::from_value::<PersistedQuery>(value.clone()).ok()
-                    })
-                    .and_then(|persisted_query| {
-                        hex::decode(persisted_query.sha256hash.as_bytes()).ok()
-                    });
-
-                let body_query = req.originating_request.body().query.clone();
-
-                match (maybe_query_hash, body_query) {
-                    (Some(query_hash), Some(query)) => {
-                        if query_matches_hash(query.as_str(), query_hash.as_slice()) {
-                            tracing::trace!("apq: cache insert");
-                            let _ = req.context.insert("persisted_query_hit", false);
-                            cache.insert(query_hash, query);
-                        } else {
-                            tracing::warn!(
-                                "apq: graphql request doesn't match provided sha256Hash"
-                            );
-                        }
-                        Ok(ControlFlow::Continue(req))
+        let service = self.inner.clone();
+        Box::pin(async move {
```

Comment:
I don't understand why we aren't using the `AsyncCheckpointService` now.

Reply:
I don't remember why I did not use it in the end, I'll try again and see if it fits

Reply:
right, it works: 8cd62fe6